notification/slack/usergrouperror.go

```diff
@@ -0,0 +1,78 @@
+package slack
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+	"text/template"
+
+	"github.com/google/uuid"
+	"github.com/slack-go/slack/slackutilsx"
+	"github.com/target/goalert/notification"
+)
+
+// userGroupError has information about an error that occurred while updating a user-group.
+type userGroupError struct {
+	ErrorID      uuid.UUID
+	GroupID      string
+	ScheduleID   string
+	ScheduleName string
+	Missing      []notification.User
+
+	OnCall string
+
+	callbackFunc func(string, ...url.Values) string
+}
+
+// ErrorRef returns a string that can be used to reference the error in a Slack message.
+func (e userGroupError) ErrorRef() string {
+	return fmt.Sprintf("`SlackUGErrorID=%s`", e.ErrorID.String())
+}
+
+// GroupRef returns a string that can be used to reference the user-group in a Slack message.
+func (e userGroupError) GroupRef() string {
+	return fmt.Sprintf("<!subteam^%s>", e.GroupID)
+}
+
+// MissingUserRefs returns a string that can be used to reference the missing users in a Slack message.
+func (e userGroupError) MissingUserRefs() string {
+	var refs []string
+	for _, u := range e.Missing {
+		urlStr := e.callbackFunc(fmt.Sprintf("users/%s", url.PathEscape(u.ID)))
+		refs = append(refs, slackLink(urlStr, u.Name))
+	}
+	return strings.Join(refs, ", ")
+}
+
+// ScheduleRef returns a string that can be used to reference the schedule in a Slack message.
+func (e userGroupError) ScheduleRef() string {
+	urlStr := e.callbackFunc("schedules/" + url.PathEscape(e.ScheduleID))
```

Comment:
To promote good secure coding habits, we should never directly concatenate strings.
```suggestion
	urlStr := e.callbackFunc(fmt.Sprintf("schedules/%s", url.PathEscape(e.ScheduleID)))
```